performances/src/profiling.ts

```diff
@@ -0,0 +1,159 @@
+  const client = await page.target().createCDPSession()
+  const stopCPUProfiling = await startCPUProfiling(options, client)
+  const { stopMemoryProfiling, takeMemoryMeasurements } = await startMemoryProfiling(options, client)
+  const stopNetworkProfiling = await startNetworkProfiling(options, client)
+
+  return {
+    takeMeasurements: async () => {
+      await takeMemoryMeasurements()
+    },
+    stopProfiling: async (): Promise<ProfilingResults> => ({
+      memory: await stopMemoryProfiling(),
+      cpu: await stopCPUProfiling(),
+      ...stopNetworkProfiling(),
+    }),
+  }
+}
+
+async function startCPUProfiling(options: ProfilingOptions, client: CDPSession) {
+  await client.send('Profiler.enable')
+  await client.send('Profiler.start')
+
+  return async () => {
+    const { profile } = await client.send('Profiler.stop')
+
+    const timeDeltaForNodeId = new Map<number, number>()
+
+    for (let index = 0; index < profile.samples!.length; index += 1) {
+      const nodeId = profile.samples![index]
+      timeDeltaForNodeId.set(nodeId, (timeDeltaForNodeId.get(nodeId) || 0) + profile.timeDeltas![index])
+    }
+
+    let totalConsumption = 0
+    let sdkConsumption = 0
+    for (const node of profile.nodes) {
+      const consumption = timeDeltaForNodeId.get(node.id) || 0
+      totalConsumption += consumption
+      if (isSdkUrl(options, node.callFrame.url)) {
+        sdkConsumption += consumption
+      }
+    }
+
+    return { total: totalConsumption, sdk: sdkConsumption }
+  }
+}
+
+async function startMemoryProfiling(options: ProfilingOptions, client: CDPSession) {
+  await client.send('HeapProfiler.enable')
+  await client.send('HeapProfiler.startSampling', {
+    // Set a low sampling interval to have more precise measurement
+    samplingInterval: 100,
+  })
+
+  const measurements: Array<{ sdkConsumption: number; totalConsumption: number }> = []
+
+  return {
+    takeMemoryMeasurements: async () => {
+      await client.send('HeapProfiler.collectGarbage')
+      const { profile } = await client.send('HeapProfiler.getSamplingProfile')
+
+      const sizeForNodeId = new Map<number, number>()
+
+      for (const sample of profile.samples) {
+        sizeForNodeId.set(sample.nodeId, (sizeForNodeId.get(sample.nodeId) || 0) + sample.size)
+      }
+
+      let totalConsumption = 0
+      let sdkConsumption = 0
+      for (const node of iterNodes(profile.head)) {
+        const consumption = sizeForNodeId.get(node.id) || 0
+        totalConsumption += consumption
+        if (isSdkUrl(options, node.callFrame.url)) {
+          sdkConsumption += consumption
+        }
+      }
+      measurements.push({ totalConsumption, sdkConsumption })
+    },
+
+    stopMemoryProfiling: async () => {
+      await client.send('HeapProfiler.stopSampling')
+
+      measurements.sort((a, b) => a.sdkConsumption - b.sdkConsumption)
+      const { sdkConsumption, totalConsumption } = measurements[Math.floor(measurements.length / 2)]
```

Comment:
FMU here we are sorting the measurements and taking the one in the middle. This means that if for example the measurements are:
```
[
  { sdkConsumption: 1, totalConsumption: 100 },
  { sdkConsumption: 2, totalConsumption: 100 },
  { sdkConsumption: 50, totalConsumption: 100 },
]
```
We will return `{ sdkConsumption: 2, totalConsumption: 100 }`.
This doesn't fail right to me, maybe computing the average will make more sense.

Reply:
Yes this is how we compute a median value! In performance analysis I felt like we tend to prefer medians over averages, so we can compare with various percentiles. But I may be wrong here, no strong opinion on the subject.

Reply:
I'd also choose median to avoid being too sensitive to outliers (if any), since it is mentioned on the report I'd say it fine